nixos/services/postgresql.nix

```diff
@@ -63,205 +97,304 @@ in {
+        postgresqlPkg.pkgs.periods
+        postgresqlPkg.pkgs.postgis
+        (pkgs.rum.override { postgresql = postgresqlPkg; })
+      ];
+
+    in {
+
+      systemd.services.postgresql.unitConfig = {
+        ConditionPathExists = [
+          # There is an upgrade running currently, postgresql must not start now.
+          "!${upstreamCfg.dataDir}/fcio_stopper"
+        ];
+      };
 
-    extensions = lib.optionals (lib.versionOlder cfg.majorVersion "12") [
-      (pkgs.postgis_2_5.override { postgresql = postgresqlPkg; })
-      (pkgs.temporal_tables.override { postgresql = postgresqlPkg; })
-      (pkgs.rum.override { postgresql = postgresqlPkg; })
-    ] ++ lib.optionals (lib.versionAtLeast cfg.majorVersion "12") [
-      postgresqlPkg.pkgs.periods
-      postgresqlPkg.pkgs.postgis
-      (pkgs.rum.override { postgresql = postgresqlPkg; })
-    ];
+      systemd.services.postgresql.bindsTo = [ "network-addresses-ethsrv.service" ];
+
+      systemd.services.postgresql.preStart = lib.mkBefore ''
+        if [[ -e ${upstreamCfg.dataDir}/fcio_migrated_to ]]; then
+          echo "Error: cannot start because the migration marker ${upstreamCfg.dataDir}/fcio_migrated_to is present."
+          echo "This data dir must not be used anymore if there is a newer data dir."
+          echo "Maybe the wrong postgresql role version is selected?"
+          echo "See 'sudo -u postgres fc-postgresql list-versions'"
+          echo "and 'sudo -u postgres cat ${upstreamCfg.dataDir}/fcio_migrated_to.log'"
+          echo "Only delete the marker if you sure that you want to use this data dir."
+          exit 2
+        fi
+
+        if [[ -e ${upstreamCfg.dataDir}/fcio_upgrade_prepared ]]; then
+          echo "Error: cannot start because the upgrade preparation marker ${upstreamCfg.dataDir}/fcio_upgrade_prepared is present."
+          echo "Is there an unfinished manual upgrade or did you mean to enable autoupgrade?"
+          echo "See 'sudo -u postgres fc-postgresql list-versions'"
+          echo "and 'sudo -u postgres cat ${upstreamCfg.dataDir}/fcio_upgrade_prepared'"
+          echo "Only delete the marker if you sure that you want to use this data dir."
+          exit 2
+        fi
+      '';
 
-  in {
+      systemd.services.postgresql.postStart =
+      let
+        psql = "${postgresqlPkg}/bin/psql --port=${toString upstreamCfg.port}";
+      in ''
+        if ! ${psql} -c '\du' template1 | grep -q '^ *nagios *|'; then
+          ${psql} -c 'CREATE ROLE nagios NOSUPERUSER NOCREATEDB NOCREATEROLE NOINHERIT LOGIN' template1
+        fi
+        if ! ${psql} -l | grep -q '^ *nagios *|'; then
+          ${postgresqlPkg}/bin/createdb --port ${toString upstreamCfg.port} nagios
+        fi
+        ${psql} -q -d nagios -c 'REVOKE ALL ON SCHEMA public FROM PUBLIC CASCADE;'
+
+        ln -sfT ${postgresqlPkg} ${upstreamCfg.dataDir}/package
+        ln -sfT ${postgresqlPkg}/package /nix/var/nix/gcroots/per-user/postgres/package_${cfg.majorVersion}
```

Comment:
???

Reply:
Also, are these links ever going to be cleared up?

Reply:
link in 160 is fixed now and points to the package link in the data directory. Both are cleaned up by the upgrade command at the end of the process.